Enable automatic release build via GitHub Actions

dannysauer opened the pull request:
Set up automatic release generation on tag creation

## Why is this PR needed?

Add a GitHub action which automatically creates a release upon pushing a `v*` tag

## Anything else a reviewer needs to know?

This is basically the heart of ci/packaging/suse/changelog_maker.sh pulled out into a Github Actions job

## Info for QA

N/A; does not directly affect skuba itself

### Related info

Refs:
 * SUSE/avant-garde#1950

## Docs

Developer docs (Confluence) should be updated.

```bash
git tag v42.19 -s -m 'bump skuba minor version'
git push --tags
make suse-release
```
The last step needs to be automated in Jenkins next (since GitHub sadly can't initiate the build)

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

<!-- Remember, if this is a work in progress please pre-append [WIP] to the title until you are ready! 
    If you can, please apply all applicable labels to help reviews out! -->

dannysauer commented:
It might make sense to alter the `changelog_maker.sh` script to have a "git" command line option (or to detect an environment variable), so the changelog formatting could all be kept in one place.  Then the action could simply run the same script instead of partially duplicating the logic.

Itxaka commented on .github/workflows/create_release.yaml:
```diff
@@ -0,0 +1,53 @@
+          PREV_TAG=$(git describe --abbrev=0 --tags "${{ github.ref }}^")
+          echo "::set-env name=baseRef::$PREV_TAG"
+
+      - name: Generate changelog
```
cant we use an existing action that generates the changelog?

https://github.com/marketplace/actions/generate-changelog

dannysauer replied:
I skipped that option in general because it delegates the formatting to a third party. It's pretty easy to generate the content ourselves, and this gives us full control over the message appearance.

Though, I honestly didn't look at all the options out there. That specific one is surprisingly configurable and does some pretty neat stuff, so it might be worth reconsidering if we really want to use the same content here as in OBS. :)

dannysauer replied:
Changelog (actually _changelogs_) generation moved to external script.

Itxaka reviewed (commented):
just one comment, looks good to me

dannysauer commented:
Updated.  This now generates all of the files needed by OBS upon creation of the release tag.  So, to create a release in OBS, all that's necessary now is to fetch the obs.tar.gz from the corresponding release, prepend the changefile additions (argh), and go forward.  If the changelog prepending part could be done with a service, this would suddenly be really easy.

dannysauer commented:
This is all behaving properly with the external scripts now.

Example Action log at https://github.com/dannysauer/skuba/actions/runs/296786345
Example Release at https://github.com/dannysauer/skuba/releases/tag/v42.38

I changed your older scripts around a bit more than originally planned, @davidcassany, but to make up for that I added in an [environment dump](https://github.com/SUSE/skuba/pull/1371/files#diff-915a5e2154f88df385815fe3b9883e7dR26-R28) in the job just for you. ;)

The shell lint check fails because the linter is broken.  It's perfectly legitimate to use a multi-line command inside the `$()` structure without line continuation characters.

jordimassaguerpla reviewed (approved):
Hi. I saw the demo and I think this is very good. I can't judge on the code, so I will leave the code revision per se to David.

davidcassany reviewed (commented):
My only comment is that I would remove the OBS interaction from the scripts here (`osc` calls) and make them happen in the follow up jenkins job that will actually do the commit of the packaged skuba. Thoughts?

So far all the rest looks good to me :+1: Thanks!

dannysauer commented on ci/packaging/suse/rpmfiles_maker.sh:
```diff
@@ -1,40 +1,81 @@
+        mkdir -p "${rpm_files}"
+        cp "${tmp_dir}"/* "${rpm_files}"
+        log "Find files for RPM package in ${rpm_files}"
+
```
I left the local execution essentially unchanged, so it generates a directory instead of a tarball.  The only addition was to also generate the body for the GitHub release comment, making copy-paste a little easier.

dannysauer commented on ci/packaging/suse/rpmfiles_maker.sh:
```diff
@@ -1,40 +1,81 @@
+                   || echo -n '')
+        if [[ -n "$ibs_user" ]]
+        then
+            do_ibs_update "$ibs_user"
```
I could see pulling this logic into a separate script that Jenkins runs.  I wanted to leave it in script which could be run locally as well, though, so that the command-line process would work the same way as an automatic Jenkins execution.  Imagine, for example, that for some reason we wanted to release skuba with a previous tag using a plus version string or similar.

I think what I'd like to do is to handle that change in the same PR as creation of the Jenkins job, though.  It makes sense to me if we bundle all of the Jenkins-relevant changes into one "automate release with Jenkins" PR.

dannysauer replied:
With the Github Action, however, I'll note that the Jenkins flow is quite a bit simpler.  Because the Action creates a release with an artifact containg the OBS files, no checkout is necessary. All the Jenkens script needs to do is:

1. get OBS tarball from the current skuba release (see https://github.com/SUSE/skuba/pull/1371#issuecomment-701043136 above)
1. extract tarball
1. prepend changelog from tarball to current changelog
1. update non-changelog files in OBS repo direct from release tarball
1. run services to update vendored modules
1. commit
1. create SR (maybe)

I've been looking at the details of the OBS services we have, and several of those steps might be doable via _service directly; what I'd *really* like to do is be able to just hit the "run service" endpoint in IBS and have it do the download, extract, and changelog update.  But we might need to do the download and changelog update separately, and then merely let obs do the extraction & vendoring

This is more of a discussion for the Jenkinsfile PR, though.  I'll get that opened up in a sec and move this over there.

dannysauer replied:
Jenkins PR is #1395

davidcassany reviewed (approved):
LGTM
